[1단계 - 자동차 경주 구현] 카고(정민호) 미션 제출합니다.

kargowild opened the pull request:
## 리뷰받고 싶은 포인트

안녕하세요! 이번 미션에서 중점적으로 리뷰 받고 싶은 점을 정리해보았습니다.

- Car 객체 설계
`0~9` 사이의 랜덤넘버를 생성하고, 생성된 값이 `0~3` 사이 이면 `정지`, `4~9`사이 이면 `전진` 하라는 요구사항이 있었습니다.
이에 랜덤넘버로 생성되는 숫자가 자동차 엑셀을 누르는 파워라고 생각해서 자동차마다 엑셀을 갖도록 설계하였습니다.

처음에는 `Car` 객체에서 `moveForward`를 public으로 노출하고 `pushAccelerator`를 private으로 숨김으로써 해당 명령이 들어오면 자동차가 엑셀을 알아서 눌러 파워값을 확인한 후 움직일지를 결정하도록 설계하려 했으나, 이렇게 설계하면 `moveForward` 메서드에 랜덤넘버를 생성하는 로직이 포함되므로 테스트코드를 작성할 때 어려움이 있었습니다.

따라서 `Car` 객체의 `moveForward`를 실행할 때 매개변수로 `pushAccelerator`의 결과를 외부에서 주입해주도록 설계했습니다.

리뷰어님이 보셨을 때 해당 설계가 합리적인지, 아니면 더 나은 설계가 있을지 궁금합니다.

이 외에도 추가적으로 리뷰해주실 포인트가 있다면 맘껏 해주시면 감사하겠습니다.
매운 맛 리뷰 부탁드립니다!!

PgmJun committed:
docs: 기능 구현 문서 작성

PgmJun committed:
feat: 자동차 이름 입력 기능 구현

PgmJun committed:
feat: 시도 횟수 입력 기능 구현

PgmJun committed:
feat: 자동차 이동 기능 구현

PgmJun committed:
docs: 자동차 현재 위치 출력 기능 문서 추가

PgmJun committed:
feat: 자동차 이동 기능 구현(자동차 위치 출력 기능 추가)

PgmJun committed:
feat: 우승자 판별 기능 구현

PgmJun committed:
docs: 리팩토링 TODO 추가

PgmJun committed:
refactor: main 메서드 분리

외부에서 CarRacing을 시작시키는 역할과, CarRacing을 직접적으로 수행하는 역할의 분리 필요성을 느낌

PgmJun committed:
refactor: Car 도메인 로직 리팩토링

- Cars 생성자 매개변수 변경
- CarAccelerator 객체 생성
  - Car가 Accelerator 를 통해서 moveForward 하여 testable한 코드로 리팩토링
- RandomNumberGenerator util 패키지로 분리
- 가독성 방해되는 매직 넘버 제거
- 위 변경 내용에 적합하도록 테스트 코드 리팩토링

PgmJun committed:
test: Cars 도메인 테스트 작성

PgmJun committed:
refactor: 매직 리터럴, 넘버 제거

PgmJun committed:
refactor: TryCount 필드명 직관적으로 변경

amount는 갯수의 느낌이 강해서 유연한 value 네이밍 사용

PgmJun committed:
refactor: 미사용 import 제거 및 코드 라인 정리

PgmJun committed:
test: Car 객체 테스트 로직 추가 구현

- 자동차가 이동하지 않는 경우 테스트
- 엑셀 push 기능 테스트

PgmJun committed:
test: CarAccelerator 객체 테스트 로직 구현

- 엑셀 push 기능 테스트

PgmJun committed:
feat: carName이 null, blank 일 때 예외처리

PgmJun committed:
test: Car 객체 생성 테스트 작성

- 자동차 이름의 길이가 1 이상 5 이하로 주어지면 자동차가 정상적으로 생성된다
- 자동차 이름의 길이가 1 미만 5 초과로 주어지면 자동차가 정상적으로 생성되지 않는다
- 자동차 이름이 null로 주어지면 자동차가 정상적으로 생성되지 않는다
- 자동차 이름이 blank로 주어지면 자동차가 정상적으로 생성되지 않는다

PgmJun committed:
test: Cars 객체 생성 테스트 작성

- 입력받은 자동차의 개수가 2대 이상이면 객체 생성에 성공한다
- 입력받은 자동차의 개수가 1대 이하이면 객체 생성에 실패한다

PgmJun committed:
test: TryCount 객체 생성 테스트 작성

- 시도 횟수가 1 이상이면 객체 생성에 성공한다
- 시도 횟수가 1 미만이면 객체 생성에 실패한다

PgmJun committed:
feat: 자동차 이름 중복 예외 처리 기능 구현

PgmJun committed:
refactor: 검증 로직 메서드 분리

PgmJun committed:
refactor: 검증 로직 메서드 분리

PgmJun committed:
remove: 처리된 TODO 제거

PgmJun committed:
refactor: indent 2 -> 1로 축약

PgmJun committed:
remove: 미사용 Field 제거

PgmJun committed:
refactor: 코드라인 정리

kargowild committed:
test: 우승자 판독 테스트코드 수정

pci2676 commented:
>안녕하세요! 이번 미션에서 중점적으로 리뷰 받고 싶은 점을 정리해보았습니다.
>
>Car 객체 설계
> 0\~9 사이의 랜덤넘버를 생성하고, 생성된 값이 0\~3 사이 이면 정지, 4\~9사이 이면 전진 하라는 요구사항이 있었습니다.
이에 랜덤넘버로 생성되는 숫자가 자동차 엑셀을 누르는 파워라고 생각해서 자동차마다 엑셀을 갖도록 설계하였습니다.
처음에는 Car 객체에서 moveForward를 public으로 노출하고 pushAccelerator를 private으로 숨김으로써 해당 명령이 들어오면 자동차가 엑셀을 알아서 눌러 파워값을 확인한 후 움직일지를 결정하도록 설계하려 했으나, 이렇게 설계하면 moveForward 메서드에 랜덤넘버를 생성하는 로직이 포함되므로 테스트코드를 작성할 때 어려움이 있었습니다.
>
>따라서 Car 객체의 moveForward를 실행할 때 매개변수로 pushAccelerator의 결과를 외부에서 주입해주도록 설계했습니다.
>
>리뷰어님이 보셨을 때 해당 설계가 합리적인지, 아니면 더 나은 설계가 있을지 궁금합니다.


제어가 어려운 부분을 외부에서 받아오도록 설계하신것은 잘하신 부분같습니다.  
통상적인 방법으로 좋은 설계라고 보여집니다.

추가리뷰는 코드에 코멘트로 달아두겠습니다.

pci2676 commented on src/main/java/domain/Car.java:
```diff
@@ -0,0 +1,52 @@
+
+    private void validateName(String name) {
+        if (name == null || name.isBlank()) {
+            throw new IllegalArgumentException();
```
exception에 문제상황을 설명할 수 있는 문구를 추가하시면 좋을것 같습니다.

다른 부분도 동일하게 반영해주세요.

kargowild replied:
다른 부분들을 신경쓰느라 기본적인 것을 놓쳤네요..! 넵 반영하도록 하겠습니다

pci2676 commented on src/main/java/domain/Car.java:
```diff
@@ -0,0 +1,52 @@
+package domain;
+
+public class Car {
+    private String name;
```
이름은 변경되지 않을것 같아보입니다.

kargowild replied:
페어프로그래밍을 하다보니 오히려 사소한 것들을 많이 놓쳤네요. 변경되지 않을 필드들에 모두 final을 붙여주었습니다!

pci2676 commented on src/main/java/domain/Car.java:
```diff
@@ -0,0 +1,52 @@
+        }
+    }
+
+    public int pushAccelerator() {
+        return accelerator.push();
+    }
+
+    public void moveForward(int power) {
+        if (power >= MIN_MOVABLE_POWER) {
+            position++;
+        }
+    }
```
여러가지 방법이 있을것 같습니다.

1. accelerator를 moveForward의 매개변수로 넘겨주어도 될 것 같습니다.
2. accelerator를 인터페이스로 만들고 테스트용 구현체를 별도로 구현한다면 moveForward가 매개변수를 전달받지 않고 처음에 작성하셨던 것 처럼 사용할 수 있을것 같습니다.

kargowild replied:
구체적인 방법까지 제안해주셔서 감사합니다 !! 제안하신 내용에 따라 코드를 변경해보았는데, 제가 의도를 잘 이해한 것일지 조금 헷갈려요. 변경된 코드를 확인해주실 수 있을까요?! 아래 커밋 버전에서 변경하였습니다!!

refactor: Car의 움직임을 제어하는 방식 변경

pci2676 replied:
👍 

제가 말씀드린 방법중 한가지를 잘 구현해주신것 같습니다.

제가 드리는 리뷰가 모든 상황에서 정답은 아니니 하나의 방법으로 생각해주시면 좋을것 같습니다 😄

pci2676 commented on src/main/java/domain/CarRacing.java:
```diff
@@ -0,0 +1,67 @@
+    private final InputView inputView;
+    private final OutputView outputView;
+
+    private static final String CAR_NAMES_DELIMITER = ",";
```
이 객체가 view의 구체적인 요구사항을 포함하고 있는것 같습니다.

역할과 책임을 조금 더 나누어보면 좋을것 같습니다.

kargowild replied:
콘솔 입력으로 받은 자동차들의 이름을 ","를 기준으로 나누는 책임을 CarRacing -> InputView로 이전시켰습니다!

이에 테스트코드를 수정하면서 위 내용과는 별 관련이 없지만 궁금한 점이 생겼는데요...!
```java
class CarRacingTest {
                  ...
                  
@Test
    @DisplayName("Cars 생성 성공 테스트")
    void createCarsSuccess() {
        // given
        List<String> carNames = List.of("pobi", "crong", "honux");
        // when
        List<Car> cars = carRacing.createCars(carNames).getCars();
        // then
        assertThat(cars).hasSize(3);
        assertAll(
                () -> assertThat(cars.get(0).getName()).isEqualTo("pobi"),
                () -> assertThat(cars.get(1).getName()).isEqualTo("crong"),
                () -> assertThat(cars.get(2).getName()).isEqualTo("honux")
        );
    }

    @Test
    @DisplayName("Cars 생성 실패 테스트")
    void createCarsFail() {
        assertThatThrownBy(() -> {
            carRacing.createCars(List.of("", "pobi", "aabbcc,pobi"));
        }).isInstanceOf(IllegalArgumentException.class);
    }
    ...
    
} 
```

CarRacing 객체는 createCars라는 메서드를 갖고 있기에 이에 대한 생성 테스트와 실패 테스트를 기존에 작성해두었는데, 사실 생성되거나 실패되는 기준은 Cars 객체 또는 Car 객체의 생성에 달려있기에 이미 CarsTest, CarTest 라는 테스트코드에서 이미 테스트 된 내용입니다. 이런 상황에서 과연 CarRacing의 createCars라는 메서드에 대한 테스트코드는 어느 수준으로 필요한지 궁금합니다..!

pci2676 replied:
저의 생각을 말씀드리면,

위 테스트는 하지 않을것 같습니다.

그 이유는 운영환경에서 다른 객체에게 노출되어 사용되는, 진입점이 되는 메서드는 `start()`입니다.

즉, 테스트를 위해 가시성을 열어둔 코드로 보입니다. 운영환경에서 사용되지 않는 코드는 가시성을 제한해주는것이 좋습니다.

이어서 `createCars(List<String> carNames)`의 테스트가 필요한지 생각해보셨으면 합니다.

그리고 `CarRacing`의 역할에 대해 한번 생각해보셨으면 합니다.  
제가 코드를 보고 파악한 `CarRacing`의 역할은 다음과 같습니다. 

도메인 객체들인 `Car`, `Cars`, `TryCount`들의 상호작용과 입출력을 `View`에게 전달하는 역할을 하는 일종의 오케스트레이션 객체로 보입니다.

그렇다면 `createCars`의 구현 사항이 별도의 메서드로 `CarRacing`에 구현되어 있는게 맞을지 의문이 듭니다.

테스트를 통해 확인을 해야하는 중요한 코드라고 생각하신다면 별도의 객체에게 책임을 위임하여 테스트하시는 것을 고려해보았으면 합니다. 그렇게 된다면 `CarRacing` 은 해당 기능을 구현한 객체의 메서드를 호출하는 그림이 되지 않을까 싶습니다.

마지막으로 만약 `start` 내에서 확인이 되어야하는 특정 구현내용이 있다면 `start`를 통해 간접적으로 테스트 할 것 같습니다.

kargowild replied:
말씀을 듣고보니 `CarRacing`이 도메인 객체들과 `view`를 연결해주는 컨트롤러의 역할과, 비즈니스 로직(?)이 수행해야 할 것 같은 `createCars` 역할을 같이 수행하는 것처럼 보이네요. 2단계가 MVC 패턴 기반의 리팩토링인데, 순수하게 컨트롤러의 역할만을 갖는 클래스를 새로 만들어야겠어요.

그렇게 되면 `start`를 수행하기 위해서 `view`를 이용하는 것이 아닌 컨트롤러에서 넘겨준 데이터를 이용하도록 바꾸면 되니 `start` 또한 `view`와 별개로 동작시킬 수 있게 되어 테스트할 수 있는 메서드로 포함시킬 수 있겠네요!

pci2676 commented on src/main/java/domain/CarRacing.java:
```diff
@@ -0,0 +1,67 @@
+import java.util.List;
+
+
+public class CarRacing {
+    private final InputView inputView;
+    private final OutputView outputView;
```
view 구현체를 직접 의존하고 있군요.

CarRacing은 domain에 속한다고 생각하시나요?

kargowild replied:
그렇네요. 
개인적으로 MVC패턴에 대해 잘 알지는 못하는데, 제 시야에서는 Controller에 속한다고 생각합니다. 
java 패키지 밑에 racingcar라는 패키지를 새로 만들고, CarRacing 클래스를 domain이 아닌 racingcar에 속하도록 디렉토리를 변경하였습니다! 제가 이해한 내용이 맞을까요?

pci2676 commented on src/main/java/domain/CarRacing.java:
```diff
@@ -0,0 +1,67 @@
+        }
+    }
+
+    public List<String> getWinners(Cars cars) {
+        int winnerPosition = cars.getWinnerPosition();
+
+        return cars.getCars().stream()
+                .filter(car -> car.getPosition() == winnerPosition)
+                .map(Car::getName)
+                .toList();
+    }
```
이 객체가 하지 않아도 되는 일이 포함되어 있는것 같습니다.

kargowild replied:
동의합니다. 우승자를 가리는 역할은 Referee 클래스가 하도록 변경해보았습니다! 제가 의도를 잘 해석했을까요?

pci2676 replied:
Referee는 상태를 가지지 않고 static한 정적 메서드만 가지고 있는것 같네요.

상태를 가지고 있는 객체를 생성하는 건 어떨까요?

일급컬렉션도 단일책임원칙을 가지는 것이 좋아보입니다.

우승자의 정보를 얻기 위해 winnerPosition을 구하기 위해 `Accelerator`가 필요할까요?  
필요하지 않다면 응집도가 높다고 볼 수 있을까요?

여러가지 고민을 해보시고 다른방향으로 구현해보시는 것도 좋을것 같습니다 😄

kargowild replied:
다시 보니 `Cars`가 `Accelerator`를 멤버변수 가지고 있을 필요는 없어보여서 `tryMove`를 호출할 때 매개변수로 `Accelerator`를 받도록 변경함으로써 `Cars`가 `List<Car>`외에 다른 멤버변수를 갖지 않는 일급컬렉션으로 만들었습니다!

우선 `Cars`가 오직 `경주에 참여하는 자동차들을 동시에 움직이게 한다`라는 책임만 갖고 있다고 보았고, 따라서 `winnerPosition`을 구하는 책임은 `Referee`에게 이전해봤어요. 또한 `Referee`가 `Cars`에 의존할 필요는 없고 `자동차이름 : 위치` 에 해당하는 Map 정보만 알면 될 것 같아서 멤버변수로 `Cars`가 아닌 `Map<String, Integer> carsPosition`을 갖도록 설계해보았어요.

이렇게 장황하게 설명을 드린 이유는 한 가지 고민되는 포인트가 있기 때문입니다..! 
프리코스를 진행하며 `객체가 가진 상태를 getter로 가져와 외부에서 처리하지 말고, 상태를 갖고 있는 객체에게 직접 시켜라`라는 글들을 여럿 보았는데, `Referee`가 현재 하는 일이 `Cars`의 상태를 getter로 가져와 외부에서 처리하는 듯한 느낌을 받았기 때문이에요. 비밥이 보기엔 어떤가요? 현재 역할 분담이 괜찮아 보이나요??

아래 커밋에 위 내용을 반영해보았는데, 현재 PR이 아닌 2단계 PR에 포함될 것 같아요. 2단계 리뷰때 확인 부탁드립니다!! 🙇‍♂️
refactor: winnerPosition을 구하는 역할을 Cars에서 Referee로 이전

pci2676 replied:
2단계 PR을 올려주실때 지금 반영해주신 부분을 한번 더 짚어주시겠어요?

제가 위와 같이 코멘트가 달리는것을 추적하기 어려울 수 있어서 부탁드립니다.

kargowild replied:
넵 알겠습니다!!! :)

pci2676 commented on src/main/java/domain/Cars.java:
```diff
@@ -0,0 +1,54 @@
+
+
+public class Cars {
+    private List<Car> cars;
```
cars 객체를 대체할 일이 있을까요?

kargowild replied:
대체할 일이 없을 것 같습니다! final 처리 하겠습니다!

pci2676 commented on src/main/java/domain/Cars.java:
```diff
@@ -0,0 +1,54 @@
+        validateCarNameDuplication(cars);
+    }
+
+    private void validateCarNameDuplication(List<Car> cars) {
+        int actualCarAmount = cars.size();
+        long distinctCarAmount = cars.stream()
+                .map(Car::getName)
+                .distinct()
+                .count();
+        if (actualCarAmount != distinctCarAmount) {
+            throw new IllegalArgumentException();
+        }
```
이름이 중복되면 안된다는 요구사항이 있었나요?

kargowild replied:
요구사항엔 없었지만, 페어와 의논한 결과 있으면 좋을 것 같아 기능을 추가해보았어요.

만약 중복을 허용하여 pobi, pobi, pobi 가 경주를 했고 1명의 우승자가 나왔다면 어떤 pobi가 우승한건지 우승메시지의 의미가 모호해질 것 같아서요..! 비밥의 의견은 어떠신가요?!

pci2676 replied:
제가 확인한 요구사항에는 없는데 어떤 이유로 셀프로 구분을 두신건가 했습니다. 단순한 궁금증이었을 뿐입니다 😅 

현실세계 애플리케이션을 만든다면 동명이인은 많을수 있으니 고유한 번호를 발번해주지 않을까 싶네요.  
하지만 이렇게 하는건 지금 요구사항에서 많이 벗어나니 그냥 듣고 흘려주세요~

pci2676 commented on src/main/java/domain/Cars.java:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    private void validateCarSize(List<Car> cars) {
+        if (cars.size() <= 1) {
```
1대의 자동차로는 게임을 수행할 수 없어야 하나요?

kargowild replied:
이 부분도 해석의 영역에 있다고 생각해요. 
페어와 의논한 결과 "경주"라는 목적에 맞으려면 최소한 2대 이상의 자동차는 있어야 한다고 생각했어요.

다시 생각해보니 부전승도 경주의 일부로 볼 수 있으려나요? 비밥의 의견은 어떤가요?!

pci2676 replied:
위에서 말씀드렸듯 요구사항에 별도로 없어서 궁금증에 질문드렸던 부분입니다 😄

pci2676 commented on src/main/java/domain/TryCount.java:
```diff
@@ -0,0 +1,20 @@
+    }
+
+    private void validate(int value) {
+        if (value < 1) {
```
1은 TryCount에서 의미를 가지는 것 같습니다.

https://javabom.tistory.com/28

kargowild replied:
동의합니다!! 또 놓쳤네요 ㅎㅎ.. 상수처리 추가하였습니다.

pci2676 commented on src/main/java/util/RandomNumberUtils.java:
```diff
@@ -0,0 +1,12 @@
+import java.util.Random;
+
+
+public class RandomNumberUtils {
```
Util 이름을 사용한 코드는 제거해주세요.

자동차게임에서 사용되는 부분은 전부 도메인 로직으로 간주하고 작성해보시면 좋을것 같습니다.

kargowild replied:
RandomNumberGenerator 라는 이름으로 변경하고 domain 패키지로 위치를 이동하였습니다!

그런데 한 가지 궁금한 점이 있어요. 페어와 의논할 당시엔 비록 현재 프로젝트는 CarAccelerator만이 RandomNumberUtils를 사용하지만, 랜덤넘버를 생성하는 로직은 자동차게임 로직 뿐만 아니라 이를 필요로 하는 클래스라면 누구나 사용해도 되는 범용적인 클래스라고 생각했거든요. 

그래서 RandomNumberUtils는 매개변수로 min, max라는 범용적인 이름의 파라미터를 받도록 만들고, 0~9 사이의 값을 생성하는 로직은 CarAccelerator에게 맡기도록 설계했어요. 이러한 경우에도 RandomNumberUtils라는 네이밍이 적절하지 않은지 궁금해요!!

pci2676 replied:
규모가 큰 애플리케이션이라면 충분히 고려해볼만한 사항이라고 생각이 되지만 지금의 경우에는 저의 의견은 조금 과하지 않나 싶습니다.

객체를 탐색할 때 계속해서 다른 객체를 탐색해나가야하는 경우 피로도가 올라가고 이해하기 점점 어려워질 수 있기 때문입니다.

kargowild replied:
객체 탐색에 대한 피로도는 처음 생각해봤던 것 같아요!! 새로운 기준을 알게되었네요! 앞으로 이 부분도 고려해서 설계해봐야겠어요

pci2676 commented on src/test/java/CarRacingTest.java:
```diff
@@ -0,0 +1,89 @@
+class CarRacingTest {
+    private CarRacing carRacing;
+
+    private static CarAccelerator accelerator;
+
+    @BeforeAll
+    static void initAll() {
+        accelerator = new CarAccelerator();
+    }
```
꼭 이렇게 초기화 되어야 하나요?

kargowild replied:
현재는 Cars가 Accelerator 를 받도록 코드를 변경했지만, 위 테스트 코드를 작성할 당시엔 Car 객체마다 CarAccelerator를 필드로 가졌기 때문에 생성시 필요했어요. 그런데 현재 요구사항으로는 Car마다 다른 Accelerator를 가져서 앞으로 나아갈 확률이 Car 마다 다른 상황이 아니라, 모든 Car 객체가 같은 CarAccelerator를 사용하는 상황이기에 CarAccelerator는 하나만 필요했고, 비슷한 맥락으로 CarRacingTest에서 수행하는 모든 테스트 메서드에서도 동일한 CarAccelerator를 사용하면 되기에 이런 방식을 선택했어요.

당시엔 `BeforeEach를 굳이 안써도 될 것 같은데? 그럼 BeforeAll을 써서 테스트 메서드마다 굳이 객체를 새로 만들지 말고, 한 번만 생성하자`라는 생각으로 사용했습니다. 그런데 다시 생각해보니 
```java
private static CarAccelerator accelerator = new CarAccelerator();
``` 
와 같이 작성해도 됐으려나요??

pci2676 replied:
객체를 공유해서 사용해서 공유된 객체에 부작용이 발생하지 않는 것으로 파악된다면 

저는 한번만 생성해서 다같이 공유해도 된다고 생각합니다. 

다만 매번 새롭게 생성되어야 하는 객체가 있다면 해당 객체는 테스트 코드 내에서 생성하는 것이 가독성측면에서 더 좋다고 생각합니다.

기존의 경우에 테스트 클래스가 실행될때 단 한번 초기화 되긴 합니다만, CarAccelerator의 선언과 초기화가 서로 다른곳에서 이루어지고 있어서 리뷰를 남겼었습니다 😄 

객체를 파악할대 시선을 옮기지 않고 한줄에 확인하는게 간단하고 좋지 않을까요?

kargowild replied:
동의합니다!!

추가로 @BeforeAll 에 대해 좀 더 찾아보았는데, 의존성 주입이 필요한 객체생성을 하는 경우가 아니라면 딱히 사용하지 않아도 될 것 같아요! 테스트 에노테이션들에 대해선 추가로 더 공부해보겠습니다!!

pci2676 commented on src/test/java/CarRacingTest.java:
```diff
@@ -0,0 +1,89 @@
+        accelerator = new CarAccelerator();
+    }
+
+    @BeforeEach
+    void init() {
+        this.carRacing = new CarRacing(new InputView(), new OutputView());
+    }
```
테스트에서 공유 의존성이 아니라면 매번 초기화하며 테스트간 공유하면서 사용하지 않는것을 권장드립니다.

kargowild replied:
바로 위에 말씀드렸던 것과 유사하게, 아래와 같이 필드에서 한 번 생성하라는 말로 이해하면 될까요?!
```java
private CarRacing carRacing = new CarRacing(new InputView(), new OutputView());
```

pci2676 replied:
위에 남긴 코멘트에서 어느정도 답변이 되었을것 같습니다. 😄

pci2676 commented on src/test/java/CarRacingTest.java:
```diff
@@ -0,0 +1,89 @@
+        // given
+        String carNames = "pobi,crong,honux";
+        // when
+        List<Car> cars = carRacing.createCars(carNames, accelerator).getCars();
```
테스트만을 위해 public으로 선언된것 같습니다.

가시성을 제한하고 테스트할 수 있도록 개선해보시면 좋을것 같습니다.

kargowild replied:
createCars 메서드가 CarRacing 클래스 내부적으로만 사용됨에도 불구하고 public으로 열려있는 부분을 지적하신 게 맞을까요?!

지금 다시 보니 createCars 메서드를 private로 닫고 이에 대한 테스트코드는 모두 지워도 될 것 같아보이는데 어떻게 생각하시나요?!

pci2676 commented on src/test/java/CarRacingTest.java:
```diff
@@ -0,0 +1,89 @@
+
+    @ParameterizedTest
+    @ValueSource(strings = {"", "pobi", "aabbcc,pobi"})
+    @DisplayName("자동차 생성 실패 테스트")
```
테스트 이름에 왜 실패하는지에 대해 설명해주면 더 좋을것 같습니다.

kargowild replied:
맞습니다...! 당시에 시간이 급했는지 잘 챙기지 못했네요. 

그런데 바로 위에 드린 질문처럼 createCar 메서드에 대한 테스트코드는 모두 삭제해도 되지 않을까 생각하는데 어떤가요?

pci2676 replied:
제가 드린 리뷰를 개선하시다가 자연스럽게 제거되거나 해결되는 리뷰들이 있을수 있는데요 그 경우 쿨하게 넘겨주세요~

pci2676 commented on src/test/java/CarRacingTest.java:
```diff
@@ -0,0 +1,89 @@
+
+    @ParameterizedTest
+    @ValueSource(ints = {0, -1, -30})
+    @DisplayName("시도 횟수 생성 실패 테스트")
```
이곳도 테스트이름만 보고는 테스트가 무엇을 나타내는지 알 수 없는것 같습니다.

kargowild replied:
여기도 마찬가지로 createTryCount 메서드를 private 으로 닫고 이에 대한 테스트코드를 제거할까 싶어요.

만약 필요한 테스트코드라면 `시도 횟수가 1회 이상이면 TryCount 객체 생성에 성공한다.` 라는 DisplayName을 주면 어떨까 해요!

pci2676 commented on src/test/java/CarRacingTest.java:
```diff
@@ -0,0 +1,89 @@
+                .isInstanceOf(IllegalArgumentException.class);
+    }
+    @Test
+    @DisplayName("우승자 판독 테스트")
```
테스트의 이름을 통해 테스트가 나타내고자 하는바를 명확히 나타낼 수 있도록 작명해주시는 것을 권장드립니다.

kargowild replied:
동의합니다!!
우승자 판독기능은 Referee 객체에게 이전되었으므로 RefereeTest에 해당 내용 반영하였습니다 !!

pci2676 commented on src/test/java/domain/CarAcceleratorTest.java:
```diff
@@ -0,0 +1,21 @@
+
+class CarAcceleratorTest {
+
+    @Test
+    @DisplayName("엑셀을 push하면 0~9 사이의 값만 리턴한다")
+    void return0To9WhenPushAccelerator() {
+        //given
+        CarAccelerator accelerator = new CarAccelerator();
+        //when
+        int actual = accelerator.push();
+        //then
+        assertThat(actual).isBetween(CarAccelerator.MIN_ACCEL_POWER, CarAccelerator.MAX_ACCEL_POWER);
```
굳이 하지 않아도 될 테스트로 보입니다.

결국 Random을 테스트하려고 하는 것 같습니다.

자바가 제공하는 라이브러리는 믿고 넘기셔도 될 것 같습니다.

kargowild replied:
동의합니다. CarAcceleratorTest에 해당 코드밖에 없어 파일을 제거하였습니다!

pci2676 commented on src/test/java/domain/CarTest.java:
```diff
@@ -0,0 +1,86 @@
+        accelerator = new CarAccelerator();
+    }
+
+    @ParameterizedTest
+    @ValueSource(strings = {"p", "po", "poo", "pooo", "poooo"})
+    @DisplayName("자동차 이름의 길이가 1 이상 5 이하로 주어지면 자동차가 정상적으로 생성된다")
+    void createCarSuccess(String carName) {
+        Assertions.assertThatCode(() -> new Car(carName, accelerator));
+    }
+
+    @ParameterizedTest
+    @ValueSource(strings = {"", "pooobi"})
+    @DisplayName("자동차 이름의 길이가 1 미만 5 초과로 주어지면 자동차가 정상적으로 생성되지 않는다")
+    void createCarFail(String carName) {
+        assertThatThrownBy(() -> new Car(carName, accelerator))
+                .isInstanceOf(IllegalArgumentException.class);
+    }
+
+    @Test
+    @DisplayName("자동차 이름이 null로 주어지면 자동차가 정상적으로 생성되지 않는다")
+    void createCarFailWhenInputNull() {
+        assertThatThrownBy(() -> new Car(null, accelerator))
+                .isInstanceOf(IllegalArgumentException.class);
+    }
+
+    @ParameterizedTest
+    @ValueSource(strings = {"", " ", "  ", "   "})
+    @DisplayName("자동차 이름이 blank로 주어지면 자동차가 정상적으로 생성되지 않는다")
+    void createCarFailWhenInputBlank(String carName) {
+        assertThatThrownBy(() -> new Car(carName, accelerator))
+                .isInstanceOf(IllegalArgumentException.class);
+    }
```
👍 

여기는 테스트의 이름이 테스트를 명확하게 설명해주고 있네요.

pci2676 commented on src/test/java/domain/CarTest.java:
```diff
@@ -0,0 +1,86 @@
+    }
+
+    @ParameterizedTest
+    @ValueSource(ints = {-1, 0, 1, 2, 3})
+    @DisplayName("4 미만의 값을 받으면 자동차가 이동하지 않는다")
```
이경우 경계 값만 테스트해주셔도 무방합니다.

kargowild replied:
그렇군요...! 경계 값 테스트 명심하겠습니다! 앞서 제안주신 Accelerator 인터페이스를 구현하면서, 경계값인 3만 테스트하도록 변경하였습니다 !!

pci2676 commented on src/test/java/CarRacingTest.java:
```diff
@@ -0,0 +1,89 @@
+    @DisplayName("시도 횟수 생성 성공 테스트")
+    void createTryCountSuccess(int tryAmount) {
+        //when
+        TryCount tryCount = carRacing.createTryCount(tryAmount);
```
테스트를 위해 가시성의 제한이 풀린것 같습니다.

가시성을 제한해주세요.

kargowild replied:
이 부분도 위에 답변드린 내용에 포함되는 것 같습니다!

pci2676 commented on src/test/java/domain/TryCountTest.java:
```diff
@@ -0,0 +1,24 @@
+
+class TryCountTest {
+
+    @ParameterizedTest
+    @ValueSource(ints = {1, 2, 3, 4})
+    @DisplayName("시도 횟수가 1 이상이면 객체 생성에 성공한다")
+    void createTryCountSuccess(int value) {
+        Assertions.assertThatCode(() -> new TryCount(value));
+    }
+
+    @ParameterizedTest
+    @ValueSource(ints = {-2, -1, 0})
+    @DisplayName("시도 횟수가 1 미만이면 객체 생성에 실패한다")
+    void createTryCountFail(int value) {
+        Assertions.assertThatThrownBy(() -> new TryCount(value))
+                .isInstanceOf(IllegalArgumentException.class);
+    }
```
테스트 케이스에서 경계 값만 테스트해도 충분해 보입니다.

kargowild replied:
경계값 테스트 한 번 더 명심하겠습니다!!!

pci2676 reviewed (changes requested):
반갑습니다. 🤠 

조금 더 개선해보고 고민해볼만한 부분에 리뷰를 남겨두었으니 확인해주시고 다시 요청해주세요.

kargowild committed:
refactor: 예외 발생 시 적절한 메시지 출력 기능 구현

kargowild committed:
refactor: 변경되지 않을 필드들에 final 키워드 추가

kargowild committed:
refactor: Car의 움직임을 제어하는 방식 변경

- 기존의 코드에서는 Car의 메서드인 moveForward와 pushAccelerator를 모두 public으로 열어두었습니다.
- 따라서 요구사항과 같이 자동차를 랜덤하게 움직이게 만들려면, 사용자는 moveForward 메서드의 매개변수로서 pushAccelerator 메서드의 반환값을 넘겨줘야 합니다.
- 이러한 방법은 번거롭고 직관적이지 않습니다. 또한 사용자가 pushAccelerator를 moveForward에게 매개변수로 주지 않고 int값을 직접 넘겨주는 실수를 저지를 가능성이 있습니다.
- 따라서 accelerator라는 역할을 moveForward의 매개변수로 넘겨주도록 리팩토링해 프로덕션 코드와 테스트 코드 모두가 좀 더 합리적으로 Car 객체를 활용하도록 개선하였습니다.

kargowild committed:
refactor: 자동차들의 이름을 ","로 구분하는 기능을 CarRacing -> InputView로 이전

kargowild committed:
refactor: CarRacing 클래스는 domain에 속하지 않으므로 디렉토리 위치 변경

kargowild committed:
refactor: 우승자를 결정하는 역할을 CarRacing에서 Referee로 이전

kargowild committed:
refactor: 최소 시도 횟수에 대한 상수처리

kargowild committed:
refactor: Util 이름을 사용하지 않도록 클래스명 변경

kargowild committed:
test: 이전에 디렉토리 구조를 변경한 것과 매치되도록 테스트 코드 디렉토리 변경 및 안쓰는 import 구문 처리

kargowild committed:
test: assertThatCode에 doesNotThrowAnyException 메서드 체인 추가 및 사용하지 않는 변수들 제거

kargowild committed:
test: DisplayName에 좀 더 명확한 설명이 들어가도록 변경

kargowild committed:
test: 자동차가 움직이지 않길 의도하는 상수의 이름을 적절하게 수정

kargowild commented:
> 반갑습니다. 🤠
> 
> 조금 더 개선해보고 고민해볼만한 부분에 리뷰를 남겨두었으니 확인해주시고 다시 요청해주세요.

반갑습니다!! ㅎㅎ
제안해주신 Accelerator 인터페이스를 적용하면서 코드가 변경된 부분이 많은데요, 확인해주시면 감사드리겠습니다!! 🙇
좋은 주말 되세요 :)

kargowild committed:
refactor: 상수, 클래스 변수, 인스턴스 변수, 생성자, 메서드 순으로 작성 되도록 수정

kargowild committed:
refactor: TryCount의 value는 변경될 여지가 없으므로 final 키워드 추가

pci2676 reviewed (approved):
리뷰드린 내용들 잘 반영해주신것 같습니다.

코멘트에 대한 답변을 달아두었습니다. 확인하시면서 추가로 개선해볼만한 부분은 다음 단계를 진행하면서 같이 반영해주시면 좋을것 같습니다.

머지하겠습니다 🚀
